src/code/main.tsx

```diff
@@ -43,53 +43,53 @@ gFirebase.postConnect.then( (imp:FirebaseImp)=> {
       else {
         hashHistory.push('/simulations/choose');
       }
+      if (callback) { callback(); }
     };
 
     const simulationChanged = function (
       prevState: any,
       nextState: any,
       replace: any,
-      callback: Function | undefined
+      callback: () => void | undefined
     ) {
-      updateSession(nextState.params.simulationName);
-      if (callback) {
-        callback();
-      }
-    };
-
-    const onEnterSimulation = function (
-      nextState: any,
-      replace: any,
-      callback: Function | undefined
-    ) {
-      updateSession(nextState.params.simulationName);
-      if (callback) {
-        callback();
-      }
+      updateSession(nextState.params.simulationName, callback);
     };
 
     const setTeacherMode = function (
       nextState: any,
       replace: any,
-      callback: Function | undefined
+      callback: () => void | undefined
     ) {
+      const simulationName = nextState.params.simulationName;
+      if (simulationName) {
+        updateSession(simulationName);
+      }
       const simulation = simulationStore.selected;
       if (simulation) {
         simulation.setIsTeacherView(true);
-        if(callback) { callback(); }
       }
+      if(callback) { callback(); }
     };
 
     const setStudentMode = function (
       nextState: any,
       replace: any,
-      callback: Function | undefined
+      callback: () => void | undefined
     ) {
-      const simulation = simulationStore.selected;
-      if (simulation) {
-        simulation.setIsTeacherView(false);
-        if(callback) { callback(); }
+      const simulationName = nextState.params.simulationName;
+      if (simulationName) {
+        // students must wait until teacher has started simulation
+        gFirebase.waitForPathToExist(`simulations/${simulationName}`, (snapshot: any) => {
+          if (simulationName) {
```

Comment:
Didn't we already unesure `simulationName` on line 80?  Could it have gone out of scope?